Mc12345678 restore paypal incontext operation

mc12345678 opened the pull request:
Fixes #1192

drbyte commented:
👍

drbyte commented:
I find it curious that PayPal's demos use a different file, `checkout.js` instead of `paypal.js` and none of their docs mention the `paypal.js` file.

mc12345678 commented:
Have found today that api/checkout.js will work as well, but there is still an issue with implementation where in process_button, because when this is called in later processing and the token is set or when cancel is selected (causing the token to be set as well as ec_cancel to be set to 1, that the javascript code provided does not get "redisplayed".

The Web address that I have that documented it's use redirects to the current in-context documentation. Further by following the path of that content, it in part directs to a version of the checkout.js file.

Recommend an alternate location to capture the inclusion of the applicable js code or a "work-around" in process_button or elsewhere so that process_button is "smarter".

mc12345678 commented:
FYI, I have a revised version that is more in line with currently published information. Currently not in a position to force replacement of the existing pull request. Can push as is (with history) or at a later time force push a "one time" solution.) Please let me know which is preferred.

mc12345678 commented:
FYI, I have a revised version that is more in line with currently published information. Currently not in a position to force replacement of the existing pull request. Can push as is (with history) or at a later time force push a "one time" solution.) Please let me know which is preferred.

mc12345678 commented:
FYI, I have a revised version that is more in line with currently published information. Currently not in a position to force replacement of the existing pull request. Can push as is (with history) or at a later time force push a "one time" solution.) Please let me know which is preferred.

drbyte commented:
I'd recommend:

a) make a new commit which updates this PR by removing unnecessary changes and adding additional changes. 

b) then your proposed changes can be reviewed here

c) then later once everything's determined to be the correct complete solution, we can consider having you rebase the entire PR into one nice clean commit which you force-push to update this PR (which is the preferable approach, since all the separate earlier commits you made to this PR aren't very atomic; they're too granular) ;
or we could consider squashing them on our end before merging (this is less clean but is better than doing a direct merge)

But, since you said you're not able to do a "force push" "currently", for whatever reason, at least start with adding yet another commit to update this PR with your latest proposed strategy.

drbyte commented:
It's throwing a JS error: `Undefined variable: paypal` when trying to call `paypal.checkout.setup(the merchant id)`

mc12345678 commented:
Looks like was also missing the modification that I expected to exist related to the button information... Adding to and forcing to the commit momentarily. (Able to work locally with a git rather than through the github website only.)

drbyte commented:
While you're working with a local git client, do you mind rebasing all the commits into one, and then doing a force push?

mc12345678 commented:
Force push finally done... Added installation keys to offer a little bit of modification capability, Addressed PayPal's implementation which adds a button regardless of an existing image within a link such that if/when possible to display the in context window that if the visitor doesn't have javascript enabled that a button is still displayed for checkout/payment. 

Instead of duplicating the button code for every file implementation (which would also easily allow each presentation of the button to be unique), a common file was generated that implements the PayPal default settings if not fully/properly installed.

drbyte commented on includes/modules/pages/checkout_confirmation/jscript_paypal_incontext.php:
```diff
@@ -0,0 +1,21 @@
+        
+        require(DIR_WS_MODULES . 'payment/paypal/paypalwpp_incontext.php');
+?>
+<!-- merchant site HTML page --><script async="async" src="//www.paypalobjects.com/api/checkout.js"></script>
```
What's the purpose of the `<!-- merchant site HTML page -->` comment? Or did that just come from some sample code?

mc12345678 replied:
Originally was from sample code, then thought to leave to make discovery within code "easier", then thought for some unknown reason that would find itself near the actual new button that is created, but in retrospect serves no real purpose.

That said, even though I originally had the required file directly in the one jscript_ file, seems like one of the "methods" to find ZC content is to reference some type of "comment" in each "section". Kind of wondering if shouldn't add a comment at the beginning and end of the file outside the if statement to support "diagnosis" of why the page(s) don't respond as expected.

Thoughts? (ie. Remove the above comment and add two minor ones, one at file opening before <?php and one after close of if referencing in-context.)

drbyte replied:
Skip the comments IMO. 
If there's an issue loading the js files then it won't be just with this one, so sending extra text to the browser 100% of the time is unnecessary clutter.

mc12345678 replied:
I was thinking more from the standpoint that in source code review, one sees that there is this in context option, but no active code, searches and finds that their issue might be the missing the merchant ID. Once entered, then assuming JS related issues that it now "works". Somewhat trying to address the potential/likely questions as best as possible.

drbyte replied:
I completely understand your logic.

But not everyone uses PayPal, and there's no need to clutter every site with unnecessary comments. 
Someone who would trace back thru source as you described is equally likely to check the admin config settings for the module and discover the same requirement.

Additionally, while I grok that lots of people have resorted to injecting a plethora of comments into every template segment, I believe it'd be better to be able to also turn off all those comments, displaying them only optionally if a master config switch were enabled for troubleshooting purposes. But that's a change that requires editing a massive number of individual files.
And, now this discussion digresses far from the topic at hand.

mc12345678 replied:
Comment to be removed and incorporated by forced push again, unless desired to "build" on this "one" at a time? :/ (Or at least to make the same change to all of the files at once instead of something like 6 times over.) Either way will be something I will get to in the next day or two based on accessibility to the desired software/hardware combination. Please continue to provide comments, insightful.

drbyte replied:
rebased force push is ideal.

drbyte commented on includes/modules/payment/paypal/paypalwpp_incontext.php:
```diff
@@ -0,0 +1,43 @@
+?><script>
+window.paypalCheckoutReady = function () {
+  paypal.checkout.setup('<?php echo MODULE_PAYMENT_PAYPALWPP_MERCHANTID; ?>', {
+    //locale: '<?php echo ${$paypalwpp_module}->getLanguageCode('incontext'); ?>',
```
I'm not sure the `$paypalwpp_modue` variable is needed. Couldn't it just be a string directly in the `new payment()` call?

And, syntactically, shouldn't `${$paypalwpp_module}->getLanguageCode(` actually be `${$classPay}->getLanguageCode(`?

mc12345678 replied:
So, here was trying to align various code to existing usage. I need to relocate from where this was taken and also verify that on first usage (shopping_cart) that the language identifier is returned. Reason for this verification is that when first implemented, received an odd error with a section of notifier code where the issues appeared unrelated. But once the require of the payment class was made require_once, the issue went away. It was then seen as odd that ${$paypalwpp_module} actually worked. So in a way this syntactical oddity was left to in a way make a point and possibly address a separate issue. If $classPay really represents the resulting class assignment, then it seems that the corrected code should be $classPay->.

To further "explain", I had found something that worked in the heat of solving the problem and hadn't gone back to perform further potential cleanup such as this. It just doesn't seem right that in calling to declare a class that suddenly the global space now includes the variable by that modules class. Not saying it's a problem, just seemed unusual and another reason why the code wasn't presented in a more "syntactical" fashion. :)

mc12345678 replied:
Found the code snippet from which this originated...

ipn_main_hander.php:

```php
  // See if the paypalwpp module is enabled.
  if (defined('MODULE_PAYMENT_PAYPALWPP_STATUS') && MODULE_PAYMENT_PAYPALWPP_STATUS == 'True') {
    $paypalwpp_module = 'paypalwpp';
    // init the payment object
    $payment_modules = new payment($paypalwpp_module);
    // set the payment, if they're hitting us here then we know
    // the payment method selected right now.
    $_SESSION['payment'] = $paypalwpp_module;
    // check to see if we have a token sent back from PayPal.
    if (!isset($_SESSION['paypal_ec_token']) || empty($_SESSION['paypal_ec_token'])) {
      // We have not gone to PayPal's website yet in order to grab
      // a token at this time.  This will send the customer over to PayPal's
      // website to login and return a token
      $$paypalwpp_module->ec_step1();
    } else {
      // This will push on the second step of the paypal ec payment
      // module, as we already have a PayPal express checkout token
      // at this point.
      $$paypalwpp_module->ec_step2();
    }
  }
```

While the above code is a "snippet" from the entire file, $payment_modules is not used at all in the remainder of that code other than to be assigned or potentially assigned again elsewhere later in the same file.  Further, the use of $$paypalwpp_module instead of ${$paypalwpp_module} while apparently clear enough for PHP 7.0 and doesn't currently cause an issue, seems like perhaps could lead to a problem in the future if not considered, but not trying to make changes just to do so. (Like the missing semi-colon in the tpl_ec_button file in the non-replaced link line.)

drbyte replied:
My bad. I had the global mixed around in my head. (Not happy with all the global vars we use in ZC, but we're working on changing that. Separate topic.)
You're right. No need to switch to ${$classPay}.

If we decide later to clean up the string to avoid assigning another unused var, we can do that in the snippet you quoted as well, at that time.

mc12345678 replied:
So should I change the use of variables in this section? Ie instead of ${$paypalwpp_module] just use $paypalwpp? I hadn't looked into the code significantly deeper to see what all happens in the function regarding the passed variable and had just considered the code reuse as "easy" to address when/if some sort of global change was to occur.

drbyte replied:
No. I'm assuming that even though you've commented-out the `locale` part you intend to keep it. In that case leave this section as-is.

mc12345678 replied:
Chicken/egg. It was commented out in the code of paypalwpp. I haven't had a good environment in which to test it, so left it for those wishing to get more from it. Similar thing of also trying to minimize future documentation. :)

drbyte commented on includes/modules/payment/paypalwpp.php:
```diff
@@ -686,6 +686,9 @@ function install() {
     $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Use InContext Checkout?', 'MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE', 'InContext', 'PayPal now offers a newer friendlier InContext (in-page) checkout mode (Requires that you enter you ...
+    $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('InContext Button Color?', 'MODULE_PAYMENT_PAYPALWPP_INCONTEXTCOLOR', 'gold', 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled PayPal offers different co ...
+    $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('InContext Button Size?', 'MODULE_PAYMENT_PAYPALWPP_INCONTEXTSIZE', 'small', 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled PayPal offers various PayPa ...
+    $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('InContext Button Shape?', 'MODULE_PAYMENT_PAYPALWPP_INCONTEXTSHAPE', 'small', 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled PayPal the PayPal button  ...
```
I "get" why you've done this. 
I'm curious: what's the likelihood of people needing to customize this? What's your gut say? 1 in 5?, 1 in 25? or 1 in 500?

(I'm mindful that we used to offer color customization options this way, but later moved them to the language file because nobody used them, and then later even PayPal deprecated and abandoned them completely.)

mc12345678 replied:
So, this was a last minute feature add (bell or whistle in my opinion) with the thought of a few things. First could completely omit any such designations in the original file as the values provided are the PayPal defaults (at this time), but also trying to think about all of the discussions in the forum about presentation and ability to modify aspects to suit. So I went all out with configurable menu options which brought to new ZC users the option to easily locate and then modify the setting(s) for the selection(s). I feel that at the moment (though pure speculation) that talking about the 1 in 500 end users, but... also not totally thinking that the end user is the one to likely adjust/consider the option as much as the one that is installing/upgrading the site. I do know that a test site on which this was applied that the default PP gold button looks terrible, the blue works but the silver makes it almost disappear but really matches the template (but it doesn't pop).

By moving all calls to this formatting into a shared file, I was trying to eliminate any such use, but then got to thinking about the DIYers and was trying to settle on a sort of happy medium considering PP seemed to think this "feature" of at least all three options to be necessary. 

Now that said, if you're thinking to just move color to a definition or similar and keep the other two as options, I could go with that and leave a note about the current potential other "styles". It all becomes mute if the shared file is brought into the jscript_ in order to support various representations of the PayPal button.

Essentially, while I'm satisfied that there is working code and believe that ZC needs anything it can get to regain the traction lost in the last 2 years or so. Don't know if it's a culture shift (generationally related) or if there is/was some ruffled feathers, but ZC in recent e-commerce reviews doesn't seem to even show. When commenters bring it up, the response is often not positive. :(

drbyte commented:
Don't get me wrong: my comments are not criticisms. Just standard code-review questions ;)

mc12345678 commented:
> Don't get me wrong: my comments are not criticisms. Just standard code-review questions ;)

No worries on my part, they're valid questions in an attempt to minimize changes and in the case of jscript the amount of unnecessary information passed to the browser, past experience with related code, and general kitten misfortune curiosity.  

If nothing else, almost everything asked about was on my irk (with no visible/noticeable impact on operation) list, though expected more also about the separation of the button action code from the checkout.js line.  Also somewhat expected something about that the PP javascript checkout.js code is called rather than "included".  Though by review of the code it seems like it is still being periodically updated.  From an end user perspective, seems best to keep it referenced instead of included... But, one might then also say the same about other software such as jQuery. :)

drbyte commented:
> expected something about that the PP javascript checkout.js code is called rather than "included". 

No. The `checkout.js` file should definitely be called from paypal directly. It's bad practice to self-host any PayPal assets. They should always be loaded from PayPal's servers.

drbyte commented:
I do wonder though: there's a lot of copy-paste duplication in all those jscript files. I can't think of any reason for them all to be duplicates of each other. Lots of room for error there.

Why not just make them all do:

```php
<?php
    require(DIR_WS_MODULES . 'payment/paypal/paypalwpp_incontext.php');
```
... and move all the `if` conditional logic into the `require`d file?

mc12345678 commented:
Had contemplated that, but was also thinking that perhaps code reuse for other purposes might exist. The central settings in the required file with the reason to use them outside of it... 

That said and several hours on which to ponder. Putting all in one file, if someone wished to specialize one or another versions of the file, they could copy the contents of the shared file into the individual file and modify to suit.

Arm sufficiently twisted. :)

drbyte commented on includes/modules/pages/checkout_confirmation/jscript_paypal_incontext.php:
```diff
@@ -0,0 +1,27 @@
+if (isset($_SESSION['paypal_ec_token']) && !empty($_SESSION['paypal_ec_token'])
+    && isset($_SESSION['paypal_ec_markflow']) && $_SESSION['paypal_ec_markflow'] = 1) {
+
+}?>
```
Not understanding why all this is here ...

mc12345678 replied:
Non-essential if the overall concept/operation is accepted, something of an alternate path that would need to be used/taken advantage of if the provided method of "clearing" the paypal_ec_token (left the paypal_ec_markflow unchanged) is considered undesirable/unacceptable when ec_cancel=1 is a parameter of the url. 

Is primarily a remnant that can be removed for operational sake, but the interaction of all these parts is pretty extensive.

The vital part if all else is accepted is the $buttonContainer and require of the file. Otherwise thus far haven't found a way (outside of a potential init_ file or other file loaded on page load/redirect) that would allow accomplishing the same functionality already offered by ZC when pressing the confirm order button while a token is assigned to an "active" connection.

mc12345678 replied:
I meant to write something earlier about the change(s) provided, might help with some of the understanding.

Currently, process_button in paypalwpp.php only provides information related to in context process if there is no previous paypal_ec_token session data and the appropriate action/settings have been applied to allow PayPal in context to be used. In straight forward processing this is fine because either the customer visited PayPal earlier and established all of the necessary information to permit shipping verification and payment collection or they haven't yet and therefore at the checkout_confirmation page are ready to goto PayPal (in this discussion In Context) where they apply the necessary credentials to complete payment. The thing is, if at any point during this process they select the cancel option in the PayPal window, then they are redirected to the cancelurl which was set depending on the logged in status of the user before attempting purchase. Well, at the point of checkout_confirmation, if I understand correctly the user either as themselves or a temporary user is logged in and therefore the cancel url is back to the checkout_shipping page. But, there is nothing at the checkout_shipping page nor forward towards checkout_confirmation that modifies the paypal_ec_token other than potentially recreating one when the payment is again posted at the checkout_confirmation page and found to be incorrect for the current purchase. 

So, on second and subsequent trips to payment_confirmation, process_button will push nothing related to the use of in context, including if the user were to go back as far as the shopping cart and then use in context payment if then immediately delivered to the checkout_confirmation page.

So two options at that point seemed possible. Make it so that some or all of the session data was removed before getting to that point, or remove the check of the existing session data and lose the operational benefit of having/maintaining that data. From the notes in paypalwpp.php it appears that the goal is to maintain the session data, so the only thing that seemed viable was to remove the session data when it would appear as most likely to need to be rebuilt anyways (cancel). So, then the issue of the checkout.js file's involvement, well, if the session data is to prevent use of in context on the checkout_confirmation page, then it is possible that it was used in advance of that page but that page is/was to be shown and the lightbox removed. To accomplish the later part requires the checkout.js file somewhere on page, but with it inside of the form code, it never seemed to work. Once pulled out, it seemed to work though came across some issues with firefox and chrome. Chrome outright didn't like the iframe if it was hosted through https and the source site was not. Firefox was not as direct about the issue.

So, perhaps that answers the other side of the potential confusion. (not sure to which aspect the question was expected to cover)

drbyte replied:
The conditional code in this file seems unnecessary here. Particularly when there's nothing inside this nested `if`.

... and (unless I'm missing something here despite your explanation) then the outer `if` is repeated in the `require`d file anyway, so feels like it could be excluded and just left to the `require`d file instead.

drbyte commented on includes/classes/observers/class.paypalwpp_incontext.php:
```diff
@@ -0,0 +1,76 @@
+     && defined('MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE') && MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE == 'InContext' 
+     && defined('MODULE_PAYMENT_PAYPALWPP_MERCHANTID')    && MODULE_PAYMENT_PAYPALWPP_MERCHANTID    != '' 
+     && defined('MODULE_PAYMENT_PAYPALWPP_STATUS')        && MODULE_PAYMENT_PAYPALWPP_STATUS        == 'True') {
+      unset($GLOBALS[_SESSION]['paypal_ec_token']);
```
I don't understand why you and others lately are referencing `$GLOBALS[_SESSION]` when $_SESSION is a superglobal and therefore `$_SESSION` is all that's needed.

mc12345678 replied:
Comes from a comment in the php manual related to unsetting the $_SESSION variable. Of course note was made 11 years ago. Based on the ZC posted related comment, and the consideration that this file could be called within a function (even though it is normally loaded in the global space) was taking that code "requirement" into consideration. Thus far in review of other areas of the php manual it discusses the "availability" of the superglobal variables but not necessarily the impact of modifying the associated data such as unsetting it.

Basically comes down to following another's lead in the face of the statement there and here about it being a superglobal and unnecessary to refer back to the $GLOBALS variable in such a case.

drbyte commented on includes/auto_loaders/config.paypalwpp_incontext.php:
```diff
@@ -0,0 +1,27 @@
+ * @license     http://www.zen-cart.com/license/2_0.txt GNU Public License V2.0
+ * @version     $Id: config.paypalwpp_incontext.php xxxx 2016-08-20 11:15:10Z mc12345678 $
+ *
+ * Stock by Attributes 1.5.4
```
A lot of unnecessary docblock components here. 

And "Stock by Attributes" can probably be removed ;)

mc12345678 replied:
Welcome suggestion of making it shorter, what minimal information is suggested if any?

Thought I got rid of that even though I was using it as a template. :) oops.

drbyte replied:
I guess it stood out because the docblock segment is longer than the lines of code.

I'd remove `@link`. 

`@package` probably should probably be more consistent across all the files you're adding. Right now you've got `paypalwpp_incontext`, `paypalwpp`, `paymentMethod`, `page`.   

I'm inclined to suggest `paymentMethod`, or if you feel the distinction is beneficial then maybe `paypal_incontext` ... for all the files you're adding via this PR.

mc12345678 replied:
Considering that some of the files could be removed (are needed only to support paypal_incontext)  and the overall functionality would remain (payment possible through PayPal), I plan on modifying to paypal_incontext for newly added files.

Will address other inconsistencies with the docblocks as well.  Expect to be able to provide updated (and corrected) within the next 14 hours. I encourage not giving up on the review and continuing to identify other issues no matter how "small".

mc12345678 commented on includes/classes/observers/class.paypalwpp_incontext.php:
```diff
@@ -0,0 +1,76 @@
+   * Generic function that is activated when any notifier identified in the observer is called but is not found in one of the above previous specific update functions is encountered as a notifier.
+   */
+  function update(&$callingClass, $notifier, $paramsArray) {
+  global $db;
```
Agree. Will strike. Good catch and thank you.

drbyte commented on includes/modules/payment/paypal/paypalwpp_incontext.php:
```diff
@@ -0,0 +1,70 @@
+      if (zen_not_null($payMod)) {
+        $_SESSION['payment'] = $payMod;
+      } else {
+        unset($GLOBALS[_SESSION]['payment']);
```
$_SESSION

mc12345678 replied:
DOH!

drbyte commented:
> @mc12345678 " I encourage not giving up on the review and continuing to identify other issues no matter how "small". There are plenty of details to consider, don't want things to die out in the end. :) "

Agreed.

mc12345678 commented:
FYI, have already considered use of four spaces for new code. Other formatting criteria for new code that has been missed?

drbyte commented on includes/modules/payment/paypalwpp.php:
```diff
@@ -306,7 +306,7 @@ function process_button() {
       };
       </script>
-      <script src=\"//www.paypalobjects.com/api/checkout.js\" async></script>";
+      <div data-paypal-button=\"yes\" data-paypal-id=\"" . MODULE_PAYMENT_PAYPALWPP_MERCHANTID . "\"></div>"
```
Is there a missing `;` here?

mc12345678 replied:
Yes, in my original code it was on the next line with a lot of commented out code between the end of that line and the end of the one below.

Added back to the end of that line.

drbyte commented on includes/classes/observers/class.paypalwpp_incontext.php:
```diff
@@ -0,0 +1,75 @@
+  
+    if ($notifier == 'NOTIFY_HEADER_START_CHECKOUT_PAYMENT') {
+      $this->updateNotifyHeaderStartCheckoutPayment($callingClass, $notifier);
+    }
```
Upon further review it strikes me that this entire `function update()` section is redundant, since the observer system automatically fires the corresponding `updateFooBarBiz()` method whenever the `FOO_BAR_BIZ` notifier point is triggered.

mc12345678 replied:
True for ZC 1.5.3 and above, but not if applied to older ZC versions.  Written with the concept that with the thankful addition of the individual function that there is the possibility of removal of the basic update function and the code would continue to support that removal. If the goal is/was to make it only fully applicable to ZC 1.5.5 then need to change it to an auto.loaded_class.php instead of the auto_loader/observer combination.

drbyte commented on includes/classes/observers/class.paypalwpp_incontext.php:
```diff
@@ -0,0 +1,75 @@
+<?php
+
+/**
+ * Description of class.paypalwpp_incontext: This class is used to support order information related to PayPal Express In Context.  This way reduces the modifications of the various header files to nearly nothing in support of troubleshooting.
```
This description feels redundant.

(You did say "no matter how trivial" ;) )

drbyte commented on includes/modules/payment/paypalwpp.php:
```diff
@@ -686,6 +686,8 @@ function install() {
     $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, date_added) values ('PayPal Merchant ID', 'MODULE_PAYMENT_PAYPALWPP_MERCHANTID', '', 'Enter your PayPal Merchant ID here. This is used for the more user-friendly In-Context checkout mode. You can obtain this value by ...
     $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('Use InContext Checkout?', 'MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE', 'InContext', 'PayPal now offers a newer friendlier InContext (in-page) checkout mode (Requires that you enter you ...
+    $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('InContext Button Size?', 'MODULE_PAYMENT_PAYPALWPP_INCONTEXTSIZE', 'small', 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled PayPal offers various PayPa ...
+    $db->Execute("insert into " . TABLE_CONFIGURATION . " (configuration_title, configuration_key, configuration_value, configuration_description, configuration_group_id, sort_order, set_function, date_added) values ('InContext Button Shape?', 'MODULE_PAYMENT_PAYPALWPP_INCONTEXTSHAPE', 'small', 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled PayPal the PayPal button  ...
```
Seems to be an extra mention of the word `PayPal` in the 2 places where you insert `MODULE_PAYMENT_PAYPALWPP_INCONTEXTSHAPE`:

> 'When PayPal\'s newer friendlier InContext (in-page) checkout mode is enabled **PayPal** the PayPal button can take different shapes.

mc12345678 replied:
Was trying to make sure that the text did not look the same for each entry and mucked up a copy/paste. First instance of Paypal removed.

drbyte commented on includes/templates/template_default/templates/tpl_checkout_confirmation_default.php:
```diff
@@ -175,7 +175,7 @@
 
         <?php
-        echo zen_draw_form('checkout_confirmation', $formActionLink, 'post', 'id="checkout_confirmation" onsubmit="submitonce();"');
+        echo zen_draw_form('checkout_confirmation', $formActionLink, 'post', 'id="checkout_confirmation" onsubmit="submitonce();"' . (defined('MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE') && MODULE_PAYMENT_PAYPALWPP_CHECKOUTSTYLE == 'InContext' && defined('MODULE_PAYMENT_PAYPALWPP_MERCHANTID')    && MODULE_PAYMENT_PAYPALWPP_MERCHANTID    != '' && defined('MODULE_PAYMENT_PAYPALWPP_STATUS')        && M ...
```
I'm still struggling over the 2 changes in this file. They're completely PayPal-specific, and it would be preferable to leave payment-module-specific stuff out of the templates.

Thoughts?

mc12345678 replied:
A separate function would/could provide the desired result, but seems would have to test for the presence of that function unless going to add it to all of the existing payment modules... Is a little "selfish" for it to just be for PayPal to be able to add to the innerhtml of the tags.

mc12345678 replied:
Working on adding process_form_params as a payment class function. Adding to paypalwpp.php payment module. Should this commit include addition of it to all payment modules or leave that as separate item/issue to address?

drbyte replied:
Given that the `payment` class methods all (if the pattern is followed) fire the method in the respective module only if it's defined, there should be no need to touch any of the other modules.

mc12345678 replied:
That was the consideration being applied. Saw that at some point it seems (from visual inspection) that the other modules had the "other" functions included even if returning false. Wanted to be sure to address that open "issue" and related consistency.

mc12345678 commented on includes/modules/pages/login/jscript_paypal_incontext.php:
```diff
@@ -0,0 +1,16 @@
+ * @version $Id: Author: mc12345678  Wed Aug 24 20:20:09 2016 -0500 Modified in v1.6.0 $
+ */
+ 
+    $buttonContainer = 'ecicButton';
```
Causes a javascript error when page is accessed prior to the cart having contents. ($ec_enabled variable or similarly named not providing a false response)

Updating to identify the button the same way the template code is written or '' if not true. All within a line instead of a formal if/else construction.

mc12345678 commented:
One thing I was thinking about. In the middle of the shipping module paypalwpp.php there is a statement that as of Dec 2015 there is a need to use Mark instead of Sole for using PayPal In Context. Was trying to figure out the conditions in which that was fully necessary and if it shouldn't have additional controls to allow Sole "on occasion".

drbyte commented:
> there is a statement that as of Dec 2015 there is a need to use Mark instead of Sole for using PayPal In Context.

Background:

`SOLUTIONTYPE` is used to control the default display style of the  PayPal landing/login page:
- `Mark` = "PayPal Login" - screen opens with login prompt
- `Sole` = "Credit Cards" - screen opens with Pay By Credit Card prompt

PayPal makes more money when people use funds in their PayPal account, instead of purchasing with credit cards. So, their primary goal is to get people to create accounts, and their secondary goal is to get them to transfer PayPal funds between PayPal accounts (where they get their 3% cut just for brokering it).

Thus, they default to "Mark" mode, and in my experience with various testing scenarios they don't fully support "Sole" in all of their application "flows"/"routes". Sometimes "Sole" just breaks w/o explanation. And their tech staff never have an answer why.  I've determined that the answer is: They assume and prefer "Mark" at all times, except for few places where "Sole" has been accommodated.

At the time I wrote that comment my testing was revealing that they weren't supporting Sole (the API calls were being rejected with errors, unless changed to Mark).

drbyte commented:
> Was trying to figure out the conditions in which that was fully necessary and if it shouldn't have additional controls to allow Sole "on occasion".

There are already way too many settings to configure. I'm very reluctant to add "more controls" without a solid business case.  What "additional controls" are you thinking of?, and "on what occasions"?

mc12345678 commented:
> What "additional controls" are you thinking of?, and "on what occasions"?

I was thinking more internally than user/operator. My interpretation of the need to use mark as a `SOLUTIONTYPE` was that it was needed for allowing the purchase to go straight through or some unique condition that just wasn't followed up with in more detail; however:
> Background:

> SOLUTIONTYPE is used to control the default display style of the PayPal landing/login page:

> - Mark = "PayPal Login" - screen opens with login prompt
- Sole = "Credit Cards" - screen opens with Pay By Credit Card prompt
PayPal makes more money when people use funds in their PayPal account, instead of purchasing with credit cards. So, their primary goal is to get people to create accounts, and their secondary goal is to get them to transfer PayPal funds between PayPal accounts (where they get their 3% cut just for brokering it).

> Thus, they default to "Mark" mode, and in my experience with various testing scenarios they don't fully support "Sole" in all of their application "flows"/"routes". Sometimes "Sole" just breaks w/o explanation. And their tech staff never have an answer why. I've determined that the answer is: They assume and prefer "Mark" at all times, except for few places where "Sole" has been accommodated.

> At the time I wrote that comment my testing was revealing that they weren't supporting Sole (the API calls were being rejected with errors, unless changed to Mark).

The above clearly indicates that a transaction expected to meet a certain set of criteria could not be determined (the criteria that is) and therefore and in light of some independently observed inconsistencies so for the time being there may not be a use "sole" for this set of cart conditions and use "mark" for the rest with which one can code. 

I have no specific desired need one way or another, was trying to bring back in line with existing PP EC settings where/if able to eliminate one potential variable in why something may end up not working..

drbyte commented:
> where/if able to eliminate one potential variable in why something may end up not working

Understood.

mc12345678 commented:
Discovered on a fresh install that 1) an error detected at process_confirmation if performing checkout using paypal incontext was not handled properly (return uri was undefined) and 2) that the new button shape configuration setting did not have an accurate default value so would be "blank" during setup until chosen and then saved.

drbyte commented:
@mc12345678 it's been awhile since we had dialog here on this. 
Have you been using these changes in production anywhere?

drbyte commented:
@mc12345678  I guess I'm mainly looking for confirmation that you're satisfied that these changes are complete.

mc12345678 commented:
@drbyte, yes it has been a while.

I have been using it on my hosted site as well I have recently been contacted by two independent individuals about applying it to other versions of ZC.  Most recently with a little massaging it was applied to a ZC 1.3.9 site.  The issue encountered or needed to be overcome was clearing of the token when PayPal Express was selected by an individual not yet logged in and the transaction cancelled.  Attempting to then checkout at that point using PayPal Express caused the site to redirect to the standard checkout instead of the InContext window. It appears that something in the code path between 1.3.9 and ZC 1.5x changed that required additional action to clear the token (I assume it was the ipn_main_handler code or the applicable paypalwpp.php Step1/Step2 code that changed)

I recently fully implemented the provided code (in place of what I had previously in place) on my site and just had a purchase through PayPal. Perhaps I have a product misconfiguration, but they chose PayPal Express from the shopping_cart, completed payment through PayPal, but then had to come back/continue through the checkout_payment and checkout_confirmation pages (checkout_shipping was expected to be skipped because the content is digital (though with attributes) and therefore no shipping address to be expected).  They did not cancel the paypal incontext window (which they were presented from shopping_cart), and were not asked for paypal credentials a second time once they arrived at checkout_confirmation.

Another inquirer was following some of the posts in the ZC forum and came across this code, asked a question about it, but has not responded back on success/failure of use/implementation.

drbyte commented:
Yes, the way the redirects were done (for step1/step2) in v139 have been changed since v150.

mc12345678 commented:
> @mc12345678 I guess I'm mainly looking for confirmation that you're satisfied that these changes are complete.

Ahh, so, there is a little bit of the "rub"... There is function and there is implementation.

One suggestion made was to simply apply the javascript in a way/manner that is throughout instead of on particular pages.  Ie. method of implementation.  This in part stemmed from the application of the paypal "button" at the site to have it available on each and every page instead of on/in the checkout structure only.  The code is maleable enough to handle that, but it requires "user" understanding in order to apply (or someone to apply it for them).  The same is true if the template names a button on a given page with a different designation than provided in the default template(s). 

So... I don't have significant input on applying the code in order to consider application to non-default ZC stores as "complete".

drbyte commented:
> So... I don't have significant input on applying the code in order to consider application to non-default ZC stores as "complete".

No worries.  Those sorts of adaptations evolve over time.

drbyte commented:
In reviewing the code for this PR today I'm satisfied to merge it, unless you have any objections.

👍

mc12345678 commented:
From an "inclusion" perspective, I thought that I had (somewhere?) identified the criteria needed to implement on a site that is/was perhaps modified away from a default version, ie. in effect to know the return points, origination points, and what "data" is needed to support the overall operation.  

In a way I used two other "popular" ZC plugins as a frame of reference for implementation such that the visitor is provided the associated code/data only when it has been enabled in the admin with correct data and the specific page(s) have been identified as pages applicable to operate/process the InContext code. Those plugins have javascript files installed on the applicable page, instead of being present in a central location with a "switch" to identify whether to enable/disable or what button is in question/should be sought. Though neither of those two plugins have quite the same "need" of the code... Ie. in principle if a cancel destination doesn't have the js code, the customer could possibly end up with a "locked" page, whereas the others, if they don't have the code then an added feature isn't made available. Little different. :)

mc12345678 commented:
> In reviewing the code for this PR today I'm satisfied to merge it, unless you have any objections.

:+1:
I do not have any objections, I keep wondering what/if there is some unforeseen PayPal operation that isn't done the way that ZC has previously done it and to that I have to count on those potentially more familiar with the way of PayPal.  This in my mind provides an alternate method by which to communicate with PayPal and an improved user experience/perception.

zcwilt commented:
:+1:
